feat: improve status calculation, filtering system, pagination UX & dropdown fixes

harshvns opened the pull request:
### Summary

This PR introduces a significant set of improvements across the printer listing, filtering system, status calculation, pagination UX, and the individual printer detail page.  
It also resolves dropdown and scroll-lock issues and ensures the project builds cleanly using `yarn build` and `yarn lint` as required.

---

## Enhancements & Fixes Included

### 1. Improved Filtering System
- Added new filters:
  - Manufacturer  
  - Driver type  
  - Mechanism type  
  - Support level  
  - Color capability  
- Filters update results instantly without page reload.
- All filters work together for precise combination-based searches.
- Added **Reset Filters** button to restore defaults.
- Persisted user filter preferences using `localStorage`.

---

###  2. Dynamic Pagination (Fully Integrated)
- Added selectable page sizes: 20, 50, 100, 200
- Pagination preference is saved across sessions.
- Improved UX: users no longer need to scroll to the bottom to change the page size.
- Fixed issue where Reset did not reset pagination correctly.

Commit:
fix: status logic, filtering, pagination + dropdown fixes

harshvns commented:
Hi @rudra-iitm ,  
This PR includes the updated status calculation, enhanced filters, improved pagination UX, and fixes for dropdown/scroll-lock behavior.  
It builds and lints successfully (`yarn build` and `yarn lint` both pass).

Please review it when you get a moment.  
I'm ready to update anything required. Thank you!

Commit:
fix: remove explicit any and satisfy ESLint rules

Commit:
fix: remove explicit any and satisfy ESLint rules

harshvns commented on components/ui/simple-select.tsx:
```diff
@@ -0,0 +1,182 @@
+  const context = React.useContext(SimpleSelectContext)
+  if (!context) throw new Error("SimpleSelectItem must be used within SimpleSelect")
+
+  const isSelected = context.value === value
+
+  return (
+    <div
+      role="option"
+      aria-selected={isSelected}
+      onClick={() => context.onValueChange(value)}
+      className={cn(
+        "relative flex w-full cursor-pointer select-none items-center rounded-sm py-1.5 pl-8 pr-2 text-sm outline-none",
+        "hover:bg-accent hover:text-accent-foreground",
+        "focus:bg-accent focus:text-accent-foreground",
+        isSelected && "bg-accent/50",
+        className
+      )}
+    >
+      <span className="absolute left-2 flex h-3.5 w-3.5 items-center justify-center">
+        {isSelected && <CheckIcon className="h-4 w-4" />}
+      </span>
+      {children}
+    </div>
+  )
```
Resolved

harshvns commented on app/layout.tsx:
```diff
@@ -1,6 +1,7 @@
 import type { Metadata } from "next";
 import { Geist, Geist_Mono } from "next/font/google";
 import "./globals.css";
+import Script from "next/script";
```
Removed the unused Script import to keep the file clean. Resolved.

harshvns commented on app/page.tsx:
```diff
@@ -125,12 +278,41 @@ export default function HomePage() {
+        </div>
+      </div>
+    )
+  }, [itemsPerPage, startIndex, endIndex, filteredPrinters.length, totalPages])
```
Extracted PaginationControls into a standalone component to avoid recreating it on every render and improve performance. Resolved.

harshvns commented on app/page.tsx:
```diff
@@ -1,119 +1,272 @@
+  }, [filteredPrinters, currentPage, itemsPerPage])
+
+  const startIndex = useMemo(() => (currentPage - 1) * itemsPerPage, [currentPage, itemsPerPage])
+  const endIndex = useMemo(() => Math.min(startIndex + itemsPerPage, filteredPrinters.length), [startIndex, itemsPerPage, filteredPrinters.length])
```
Added proper conditional logic for startIndex and endIndex when itemsPerPage is -1 (Show All). Prevents negative values and incorrect ranges. Resolved.

harshvns commented on components/PrinterSearch.tsx:
```diff
@@ -1,61 +1,171 @@
-              </SelectContent>
-            </Select>
+          
+          <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 xl:grid-cols-5 gap-4">
```
Updated layout from xl:grid-cols-5 to xl:grid-cols-4 for better spacing on large screens. Resolved.

harshvns commented on app/page.tsx:
```diff
@@ -1,119 +1,272 @@
+      if (savedMechanismType) setSelectedMechanismType(savedMechanismType)
+      if (savedSupportLevel) setSelectedSupportLevel(savedSupportLevel)
+      if (savedColorCapability) setSelectedColorCapability(savedColorCapability)
+      if (savedPage) setCurrentPage(parseInt(savedPage, 10))
```
Added validation logic to ensure currentPage never exceeds totalPages after filters or items-per-page change. Resolved.

harshvns commented on lib/utils.ts:
```diff
@@ -1,6 +1,57 @@
+      : "drivers" in printer
+      ? printer.drivers?.length ?? 0
+      : 0
+  // --------------------
```
Removed the “FIXED PART” placeholder comment and replaced it with a clean, descriptive explanation of the driverCount logic. Resolved.

harshvns commented on components/PrinterPageClient.tsx:
```diff
@@ -91,281 +82,266 @@ export default function PrinterPageClient({ printerId }: PrinterPageClientProps)
 
+  // --------------------------
+  // MAIN CONTENT (SUCCESS)
+  // --------------------------
```
Simplified and standardized section comments across the file for cleaner and more consistent code style. Resolved.

harshvns commented on components/ui/simple-select.tsx:
```diff
@@ -0,0 +1,182 @@
+            "focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2",
+            "disabled:cursor-not-allowed disabled:opacity-50",
+            triggerClassName
+          )}
+          aria-expanded={isOpen}
+          aria-haspopup="listbox"
+        >
```
Added aria-label using placeholder fallback to ensure accessible naming for screen readers. Resolved.

harshvns commented on package.json:
```diff
@@ -7,7 +7,14 @@
+  "lint-staged": {
+    "*.{ts,tsx}": [
+      "eslint --fix",
+      "yarn build"
```
Removed yarn build from lint-staged to avoid unnecessary builds on each staged file. Only eslint --fix runs now. Resolved.

harshvns commented on .husky/pre-commit:
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env sh
+. "$(dirname -- "$0")/_/husky.sh"
+
+yarn build && yarn lint
```
Updated pre-commit hook to use npx lint-staged as recommended. Removed unnecessary full build/lint cycle. Resolved.

harshvns commented on scripts/combine-data.mjs:
```diff
@@ -207,6 +206,10 @@ async function combineData() {
+        const functionality = printer.functionality || '?';
+        const status = getFunctionalityStatus(functionality);
+        const finalStatus = driverDetails.length === 0 && status === 'Unknown' ? 'Unsupported' : status;
+        
```
Removed the extra blank line for consistency and cleaner formatting. Resolved.

harshvns commented on components/PrinterSearch.tsx:
```diff
@@ -1,61 +1,171 @@
 import { useState, useEffect } from "react"
 import { useDebounce } from "@/lib/hooks/use-debounce"
-import { Search, Filter } from "lucide-react"
+import { Search, Filter, X } from "lucide-react"
```
Removed unused Filter import to clean up the file. Resolved.

harshvns commented on components/PrinterPageClient.tsx:
```diff
@@ -91,281 +82,266 @@ export default function PrinterPageClient({ printerId }: PrinterPageClientProps)
+          </Card>
+        </div>
+
+        {/* RIGHT SECTION — DRIVERS */}
```
Updated comment style to match the rest of the codebase and maintain consistency. Resolved.

harshvns commented on app/page.tsx:
```diff
@@ -1,119 +1,272 @@
+    if (typeof window === 'undefined') return
     
-    // Reset pagination when filters change
+    localStorage.setItem(STORAGE_KEYS.SEARCH, searchQuery)
```
Resolved — the search query is now sanitized using String(searchQuery).trim() before being stored in localStorage to prevent saving unvalidated input.

harshvns commented on app/page.tsx:
```diff
@@ -1,119 +1,272 @@
+    return filteredPrinters.slice(startIndex, endIndex)
+  }, [filteredPrinters, currentPage, itemsPerPage])
+
+  const startIndex = useMemo(() => (currentPage - 1) * itemsPerPage, [currentPage, itemsPerPage])
```
Resolved — updated the startIndex logic to return 0 when itemsPerPage === -1, preventing negative indices and ensuring correct pagination behavior.

harshvns commented on app/page.tsx:
```diff
@@ -213,21 +416,19 @@ export default function HomePage() {
             {totalPages > 1 && (
               <div className="mt-12">
-                {/* Results Summary */}
+                <PaginationControls />
```
Resolved — PaginationControls has been converted into a proper standalone React component instead of a useCallback wrapper. It now renders correctly without unnecessary re-renders or runtime issues.

harshvns commented on lib/utils.ts:
```diff
@@ -1,6 +1,57 @@
+): PrinterStatus {
+  const functionality =
+    printer.functionality ||
+    (printer as Printer).status?.toLowerCase()
```
Resolved — added a strict check to ensure status is a string before calling .toLowerCase(), preventing any undefined/null access issues.

harshvns commented on app/page.tsx:
```diff
@@ -1,119 +1,272 @@
+        if (selectedColorCapability === 'color') {
+          return type.includes('color') || type === 'inkjet' || type === 'laser'
+        } else if (selectedColorCapability === 'monochrome') {
+          return type === 'laser' || type === 'dot-matrix'
```
Resolved — updated the color capability filter to use a conservative string-based inference without treating “laser” as both color and monochrome. This removes ambiguity and produces consistent results.

harshvns commented:
Hi @rudra-iitm , I’ve addressed all review comments and resolved every thread. The PR is ready for workflow approval and review.

rudra-iitm commented:
The new Items per page looks better, but keeping it at the bottom is not a good idea. When I tried it, I selected Show All and then had to scroll all the way down to find it, and it even disappeared in Show All mode.
Please move it to the top (or somewhere more visible) so it’s easier to switch, and ensure that it does not disappear when Show All is selected. Also, please change the status from partial to mostly.

Commit:
feat: move Items per page selector to top and remove bottom instance

harshvns commented:
Hi @rudra-iitm — I’ve moved the “Items per page” selector to the top and removed the bottom instance. It now stays visible even when “Show All” is selected. I’ve also updated the status from “Partial” to “Mostly” across the codebase.
All requested changes are now completed.

rudra-iitm commented:
Looks good, merging it now.
